specifyweb/frontend/js_src/lib/components/WorkBench/WbValidation.tsx

```diff
@@ -15,7 +14,34 @@ import {
 import type { WbMeta } from './CellMeta';
 import type { UploadResult } from './resultsParser';
 import { resolveValidationMessage } from './resultsParser';
-import type { WbView } from './WbView';
+import type { Workbench } from './WbView';
+
+type UploadResults = {
+  ambiguousMatches: WritableArray<
+    WritableArray<{
+      physicalCols: RA<number>;
+      mappingPath: MappingPath;
+      ids: RA<number>;
+      key: string;
+    }>
+  >;
+  recordCounts: Partial<Record<Lowercase<keyof Tables>, number>>;
+  newRecords: Partial<
```

Comment:
```suggestion
type UploadResults = {
  readonly ambiguousMatches: WritableArray<
    WritableArray<{
      physicalCols: RA<number>;
      mappingPath: MappingPath;
      ids: RA<number>;
      key: string;
    }>
  >;
  readonly recordCounts: Partial<Record<Lowercase<keyof Tables>, number>>;
  readonly newRecords: Partial<
```

try to keep things as readonly as possible

reasons:
- `readonly` is a form of documentation - telling anyone reading the code what values would be mutated and what values won't be
- `readonly` makes typescript check that we don't mutate this property thus adding extra safety/avoiding bugs that come with mutating things
- `readonly` encourages more functional-style programming with immutable data structures